### PR TITLE
Add ProcessLinkerXmlBase to NativeAOT

### PR DESCRIPTION
**Pull request by tlakollo**
Add ReferenceSource Folder from linker in NativeAOT
Add ProcessLinkerXmlBase from linker and move it to use NativeAOT specifics

### CODE REVIEW

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ReferenceSource/BodySubstituterStep.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+using System.IO;
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ReferenceSource/BodySubstituterStep.cs:1]: Won't need this file.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ReferenceSource/LinkAttributesStep.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ReferenceSource/LinkAttributesStep.cs:1]: Won't need this file.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ReferenceSource/ProcessLinkerXmlStepBase.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ReferenceSource/ProcessLinkerXmlStepBase.cs:1]: Won't need this file.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ReferenceSource/ResolveFromXmlStep.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ReferenceSource/ResolveFromXmlStep.cs:1]: Won't need this file.

**Review by MichalStrehovsky — COMMENTED** [review]
Could you split the second commit into:

* Copy reformatted source from the linker repo.
* Make changes to the file so that it builds.

(I would just copy the buildable files to the side, `git reset --hard HEAD~`, copy files from the linker repo, let VS reformat it, commit that, and then copy over the buildable files you backed up, and commit that).

I want this structure so that the review can focus on the parts that are unique to NativeAOT. We don't need to review parts we're taking as-is from the linker repo.

**Comment by MichalStrehovsky** [comment]
The baseline version of FeatureSettings.cs  doesn't look right (it's just `Class1`)

**Comment by MichalStrehovsky** [comment]
Could you keep the TryConvertValue under an `#if false` block? We don't need it right now, but we will need it and at that point we'll want the structure to be the same anyway. It will make the diff look better too.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs [hunk 5]
```diff
@@ -38,33 +44,33 @@ public abstract class ProcessLinkerXmlBase
         const string AllAssembliesFullName = "*";
         protected const string XmlNamespace = "";
 
-        protected readonly string _xmlDocumentLocation;
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs]: It looks like we'll need this for error messages - why remove it?

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs [hunk 6]
```diff
@@ -38,33 +44,33 @@ public abstract class ProcessLinkerXmlBase
...
         readonly XPathNavigator _document;
-        protected readonly (EmbeddedResource Resource, AssemblyDefinition Assembly)? _resource;
-        protected readonly LinkContext _context;
+        protected readonly (ManifestResource Resource, ModuleDesc Module)? _resource;
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs:48]: I think it will be enough to keep just the ModuleDesc here. The resource itself is only used to mark it for removal. We do that in ManifestResourceBlockingPolicy.cs that does all the filtering necessary.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs [hunk 7]
```diff
@@ -92,18 +96,18 @@ protected virtual void ProcessXml(bool stripResource, bool ignoreResource)
...
                 if (_resource != null)
-                    ProcessAssembly(_resource.Value.Assembly, nav, warnOnUnresolvedTypes: true);
+                    ProcessAssembly(_resource.Value.Module, nav, warnOnUnresolvedTypes: true);
 
             }
-            catch (Exception ex) when (!(ex is LinkerFatalErrorException))
+            catch (Exception ex)
             {
-                throw new LinkerFatalErrorException(MessageContainer.CreateErrorMessage(null, DiagnosticId.ErrorProcessingXmlLocation, _xmlDocumentLocation), ex);
+                throw ex;
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs:104]: Could you leave what was here commented out? We'll probably want to revisit because telling the user about what XML we were processing sounds useful.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs [hunk 8]
```diff
@@ -182,90 +188,47 @@ protected virtual void ProcessTypes(AssemblyDefinition assembly, XPathNavigator
...
-                            var resolvedExternal = ProcessExportedType(exported, assembly, typeNav);
-                            if (resolvedExternal != null)
-                            {
-                                type = resolvedExternal;
-                                break;
-                            }
-                        }
-                    }
-                }
+                TypeDesc type = CustomAttributeTypeNameParser.GetTypeByCustomAttributeTypeName(assembly, fullname, throwIfNotFound: false);
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs:191]: This has different semantics around nested types at minimum; I used it as a shortcut, but if we want to be bug-for-bug compatible with illink, we'll need to build another emulator for whatever Cecil accepts there (it's another spot where the documentation of the illink file format is "whatever Cecil accepts" without any specification of what that is).

Could you add a TODO?

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs [hunk 9]
```diff
@@ -182,90 +188,47 @@ protected virtual void ProcessTypes(AssemblyDefinition assembly, XPathNavigator
                         continue;
                 }
 
-                TypeDefinition type = assembly.MainModule.GetType(fullname);
-
-                if (type == null && assembly.MainModule.HasExportedTypes)
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs]: Could you add a TODO that we might need a `ProcessExportedType` equivalent (we will need to mark the type forwarder if this went through a type forwarder).

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs [hunk 10]
```diff
@@ -182,90 +188,47 @@ protected virtual void ProcessTypes(AssemblyDefinition assembly, XPathNavigator
...
             }
         }
 
-        protected virtual TypeDefinition? ProcessExportedType(ExportedType exported, AssemblyDefinition assembly, XPathNavigator nav) => exported.Resolve();
-
-        void MatchType(TypeDefinition type, Regex regex, XPathNavigator nav)
+        void MatchType(TypeDesc type, Regex regex, XPathNavigator nav)
         {
-            if (regex.Match(type.FullName).Success)
+            if (regex.Match(type.GetDisplayName()).Success)
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs:206]: Use the CecilTypeNameFormatter to format a name instead - there's differences for nested types.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs [hunk 11]
```diff
@@ -182,90 +188,47 @@ protected virtual void ProcessTypes(AssemblyDefinition assembly, XPathNavigator
...
         {
             Regex regex = new Regex(fullname.Replace(".", @"\.").Replace("*", "(.*)"));
 
-            foreach (TypeDefinition type in assembly.MainModule.Types)
-            {
+            foreach (TypeDesc type in assembly.GetAllTypes())
                 MatchType(type, regex, nav);
-            }
-
-            if (assembly.MainModule.HasExportedTypes)
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs]: Can you add a TODO that we need to look at exported types as well? GetAllTypes won't include exported types.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs [hunk 12]
```diff
@@ -275,15 +238,15 @@ void ProcessSelectedFields(XPathNavigator nav, TypeDefinition type)
...
         }
 
-        protected virtual void ProcessField(TypeDefinition type, XPathNavigator nav)
+        protected virtual void ProcessField(TypeDesc type, XPathNavigator nav)
         {
             string signature = GetSignature(nav);
             if (!String.IsNullOrEmpty(signature))
             {
-                FieldDefinition? field = GetField(type, signature);
+                FieldDesc field = type.GetField(signature);
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs:246]: `signature` is in the form of `System.Int32 field1`. TypeDesc.GetField accepts the field name. This won't work - could you keep the `GetField` method that ILLinker has and tweak it to work with the type system? Use the Cecil name formatter to format the field type name.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs [hunk 13]
```diff
@@ -293,41 +256,19 @@ protected virtual void ProcessField(TypeDefinition type, XPathNavigator nav)
...
-                        if (field.Name == name)
-                        {
-                            foundMatch = true;
-                            ProcessField(type, field, nav);
-                        }
+                        ProcessField(type, field, nav);
                     }
                 }
-
-                if (!foundMatch)
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs]: Why delete this? We'll want to be warning compatible eventually.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs [hunk 14]
```diff
@@ -337,169 +278,61 @@ void ProcessSelectedMethods(XPathNavigator nav, TypeDefinition type, object? cus
...
             {
-                MethodDefinition? method = GetMethod(type, signature);
-                if (method == null)
-                {
-                    LogWarning(nav, DiagnosticId.XmlCouldNotFindMethodOnType, signature, type.GetDisplayName());
-                    return;
-                }
-
-                ProcessMethod(type, method, nav, customData);
+                foreach (MethodDesc meth in type.GetMethods())
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs:286]: Why the change here from `MethodDefinition? method = GetMethod(type, signature);` to a `foreach`?

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs [hunk 15]
```diff
@@ -337,169 +278,61 @@ void ProcessSelectedMethods(XPathNavigator nav, TypeDefinition type, object? cus
...
-                        if (name == method.Name)
-                        {
-                            foundMatch = true;
-                            ProcessMethod(type, method, nav, customData);
-                        }
+                        ProcessMethod(type, method, nav, customData);
                     }
                 }
-
-                if (!foundMatch)
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs]: Leave warning commented out?

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs [hunk 16]
```diff
@@ -337,169 +278,61 @@ void ProcessSelectedMethods(XPathNavigator nav, TypeDefinition type, object? cus
...
-                }
             }
         }
 
-        protected virtual MethodDefinition? GetMethod(TypeDefinition type, string signature) => null;
-
-        protected virtual void ProcessMethod(TypeDefinition type, MethodDefinition method, XPathNavigator nav, object? customData) { }
+        protected virtual MethodDesc? GetMethod(TypeDesc type, string signature) => null;
 
-        void ProcessSelectedEvents(XPathNavigator nav, TypeDefinition type, object? customData)
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs]: Why delete these?

Github's differ gets pretty confused at this point.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs [hunk 17]
```diff
@@ -522,141 +355,193 @@ protected static string GetAttribute(XPathNavigator nav, string attribute)
             return nav.GetAttribute(attribute, XmlNamespace);
         }
 
-        protected MessageOrigin GetMessageOriginForPosition(XPathNavigator position)
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs]: This one and LogWarning look useful too.

**Review by MichalStrehovsky — COMMENTED** [review]
Looks good otherwise!

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs (hunk 16 above)
Comment (tlakollo) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs]: That's a good question I know that NativeAOT uses the tables just like the ILtrim project, but there is no notion of EventDesc or PropertyDesc. My guess is that everything is represented as a method(?). Looking at the current implementation a type seems to only be able to have attributes and methods

Comment (tlakollo) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs]: It might even affect the schema PR https://github.com/dotnet/linker/pull/2500

Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs]: There is a PropertyPseudoDesc if really needed but this can be ifdeffed out for now.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs [hunk 18]
```diff
@@ -1,316 +1,510 @@
...
+        const string SignatureAttributeName = "signature";
+        const string NameAttributeName = "name";
+        const string FieldElementName = "field";
+        const string MethodElementName = "method";
+        const string EventElementName = "event";
+        const string PropertyElementName = "property";
+        const string AllAssembliesFullName = "*";
+        protected const string XmlNamespace = "";
+
+        // protected readonly string _xmlDocumentLocation;
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs:47]: Could you uncomment this and populate it from the constructors? It would be good to have the API surface already enforce providing this value.

If it complains about this being never read, add a dummy `_ = _xmlDocumentLocation` next to one of the commented out throws.

**Review by MichalStrehovsky — APPROVED** [review]
Looks good otherwise! Thank you!
